sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_arg.asciidoc

```diff
@@ -179,12 +175,17 @@ separate extensions.
 The section below describes the constructors and member functions for
 `annotated_arg`.
 
+NOTE: The template parameter `T` in the definition of `annotated_arg` template
+below must be a device copyable type or a legal parameter type as defined by the
+SYCL specification.
```

Comment:
> For example, a struct with three members might be passed as three separate kernel arguments, depending on the types of the members. It's unclear how we can preserve the annotations in these cases.

To combat this I am stating that the properties be applied on top level only and not on struct members. We are trying to see if we can issue an error if `annotated_*` classes are nested inside other structures. If we can do that I will update the spec to state that an error will be issued.

> As a first step, can we restrict the allowable types T to just the scalar types?

We do not want to restrict `annotated_arg` to scalar types. We purposefully allowed `T` to be a pointer type for users who want to annotate pointers with kernel interface properties only, i.e., the properties that are not meaningful inside a kernel body and hence the user may want to use `annotated_arg` rather than `annotated_ptr`.

> We should also make it clear that annotated_arg can only be used on the value that is actually passed as a kernel argument. You cannot, for example, use it on one member of an aggregate kernel argument.

I have provided an example in the 'Introduction' section (sub-section of 'Specifications') that states that only top-most level kernel arguments marked as `annotated_arg` will be processed for applying properties.

I am considering stating that it is undefined behavior to use `annotated_*` classes as members of structs that become kernel arguments, example:
```
struct MyType {
  annotated_arg<int, ...> arg1;
}

annotated_arg<MyType, ....> my_arg;
...
q.submit([=]{
   //  use my_arg
});
```
Currently I give a similar example and state that only the properties applied on `my_arg` (top-most level) will be processed. I want to instead state: `This will lead to undefined behavior.` Thoughts?

Reply:
> We do not want to restrict annotated_arg to scalar types. We purposefully allowed T to be a pointer type for users who want to annotate pointers with kernel interface properties only, i.e., the properties that are not meaningful inside a kernel body and hence the user may want to use annotated_arg rather than annotated_ptr.

Pointers are scalar types, so this would still be allowed.

> To combat this I am stating that the properties be applied on top level only and not on struct members. We are trying to see if we can issue an error if annotated_* classes are nested inside other structures. If we can do that I will update the spec to state that an error will be issued.

This does not address my concern.  Here is an example of a case that I do *not* want to support:

```
struct Foo {
  int *ptr1;
  int *ptr2;
  float f;
};

annotated_arg<Foo, ....> my_arg;
```

I have at least a couple of concerns.

First, this is harder to implement because the compiler may decompose the members  of `Foo` into three separate arguments, so that `ptr1`, `ptr2`, and `f` are each passed as their own argument.  If we allows a case like this, I presume the implementation needs to copy the annotations to each of those arguments.

Second, the semantics aren't clear in a case like this.  If `restrict` is applied to a structure like this, is the programmer asserting that `ptr1` and `ptr2` don't alias with each other, or is the programmer merely asserting that `ptr1` doesn't alias with any other kernel arguments that are pointers (and that `ptr2` doesn't alias with those other kernel argument pointers either)?  What if the structure has the `buffer_location<id>` property?  Does this apply to both `ptr1` and `ptr2`?

> Currently I give a similar example and state that only the properties applied on my_arg (top-most level) will be processed. I want to instead state: This will lead to undefined behavior. Thoughts?

This will be a moot point in my proposal because the code you show would be illegal.

Reply:
For, the case you listed:
```
struct Foo {
  int *ptr1;
  int *ptr2;
  float f;
};

annotated_arg<Foo, ....> my_arg;
```
kernel argument be a pointer to a struct of type `Foo` because we are planning on marking `annotated_arg`/`annotated_ptr` with `sycl_special_class` for the initial release. All sycl special classes are decomposed based on the `__init` method and not on the basis of the members. I confirmed this just this morning with Elizabeth but I will run a test to double check this.

Reply:
I don't see how that can work in the general case.  What if the structure contains a member that *must* be passed as a separate kernel argument like this:

```
struct Foo {
  accessor<int> acc;
  int *ptr1;
  int *ptr2;
  float f;
};
```

Do we have any uses cases where it makes sense to annotated a non-scalar kernel argument?  Most of the properties I see in the specification are specific to kernel arguments that are pointers.  For the few others, it's not clear to me that they make sense for a kernel argument that is a structure passed by value.

Reply:
For FPGA programs, there are important use-cases where we want to specify things like `conduit`, `register_map`, and `stable` on non pointer struct types.
Front end team is working on a change right now to stop decomposition for cases where the structs have pointers. This is expected to complete in tandem with these `annotated_*` changes or sooner.

Coming back to cases where the kernel arg structs have special classes:
Sub-case 1: accessors, samplers, streams, spec constants and half-type as kernel arg struct members 
Sub-case 2: annotated_arg, annotated_ptr as kernel arg struct members 
- this problem still exists. Users will get `struct Foo` as kernel argument instead of `accessors, int*, int*, float` in your example.

For sub-case 1 I propose that since these cases are well defined, we can state that they are not supported (we will figure out how to iron this out in the experimental stage).
For sub-case 2 I propose that we state it is undefined behavior - because these classes are meant to annotated kernel arguments and not members of kernel argument structs. We can additionally try and error out even since this is not how the class is intended to be used.

Reply:
Discussed this offline with Greg. For now we are aligned that we will support annotating user defined structs subject to certain constraints. Added a line about these constraints.